csrc/val_graph.h

```diff
@@ -0,0 +1,215 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <disjoint_set.h>
+#include <ir/all_nodes.h>
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace nvfuser {
+
+using ValGroup = std::shared_ptr<VectorOfUniqueEntries<Val*>>;
+using ValGroups = VectorOfUniqueEntries<ValGroup>;
+using ExprGroup = std::shared_ptr<VectorOfUniqueEntries<Expr*>>;
+using ExprGroups = VectorOfUniqueEntries<ExprGroup>;
+
+// Given a fusion as shown below:
+//
+// T1 = set(T0);
+// T2 = set(T1);
+//
+// T0: root [I0, I1], leaf [I0, I1]
+// T1: root [I2, I3], leaf [I2*I3/4, 4]
+// T2: root [I4, I5], leaf [I4*I5/4, 4]
+//
+// The Exact graph consists of ValGroups of:
+//
+// - {I0, I2, I4}
+// - {I1, I3, I5}
+// - {I2*I3, I4*I5}
+// - {I2*I3/4, I4*I5/4}
+// - {4, 4}
+//
+// And ExprGroups of:
+//
+// - {merge of I2 and I3, merge of I4 and I5}
+// - {split of I2*I3, split of I4*I5}
+
+class ValGraph {
+ public:
+  ValGraph() = default;
+
+  ValGraph(const ValGraph& other);
+  ValGraph(ValGraph&& other) = default;
+
+  ValGraph& operator=(const ValGraph& other);
+  ValGraph& operator=(ValGraph&& other) = default;
+
+  ValGraph(bool propagate_through_exprs)
+      : propagate_through_exprs_(propagate_through_exprs) {}
+
+  // Returns the disjoint val set.
+  const DisjointSets<Val*>& disjointValSets() const {
+    return disjoint_vals_;
+  }
+
+  // Returns the disjoint Expr set.
+  const DisjointSets<Expr*>& disjointExprSets() const {
+    return disjoint_exprs_;
+  }
+
+  // Return if there's a group entry in the graph for this expr
+  bool hasGroup(Expr* expr) const;
+
+  // Return if there's a group entry in the graph for this val
+  bool hasGroup(Val* val) const;
+
+  // Convert expr to its exprGroup, assert that it exists.
+  const ExprGroup& toGroup(Expr* expr) const;
+
+  // Convert iter domain to its IdGroup, assert that it exists.
+  const ValGroup& toGroup(Val* val) const;
+
+  // Return output/input iter domain groups of provided expr
+  // Note that the same IdGroup can show up multiple times, so the
```

Comment:
Looks like the comments is not updated yet.

Reply:
Updated